tidy3d/components/data/monitor_data.py

```diff
@@ -374,6 +420,103 @@ def dot(
+        fields_other = field_data._tangential_fields  # pylint:disable=protected-access
+
+        # Cell centers used for current tangential fields
+        bounds = self._plane_grid_boundaries
+        centers = [(bs[1:] + bs[:-1]) / 2 for bs in bounds]
+
+        # Interpolate fields from other data to cell centers
+        kwargs = {"bounds_error": False, "fill_value": 0.0}
+        interp_dict = {}
+        for dim, cents in zip(tan_dims, centers):
+            if cents.size > 0:
+                interp_dict[dim] = cents
+        fields_other = {
+            key: val.interp(kwargs=kwargs, **interp_dict) for key, val in fields_other.items()
+        }
+
+        # Tangential field component names
+        dim1, dim2 = self._tangential_dims
+        e_1 = "E" + dim1
+        e_2 = "E" + dim2
+        h_1 = "H" + dim1
+        h_2 = "H" + dim2
+
+        # Prepare array with proper dimensions for the dot product data
+        arrays = (fields_self[e_1], fields_other[e_1])
+        coords = (arrays[0].coords, arrays[1].coords)
+        # Common frequencies to both data arrays
+        f = np.array(sorted(set(coords[0]["f"].values).intersection(coords[1]["f"].values)))
+        mode_index_0 = coords[0]["mode_index"].values
+        mode_index_1 = coords[1]["mode_index"].values
+        dtype = np.promote_types(arrays[0].dtype, arrays[1].dtype)
+        dot = np.empty((f.size, mode_index_0.size, mode_index_1.size), dtype=dtype)
+
+        # Calculate overlap for each common frequency and each mode pair
+        for i, freq in enumerate(f):
+            for mi0 in mode_index_0:
+                e_self_1 = fields_self[e_1].sel(f=freq, mode_index=mi0, drop=True)
+                e_self_2 = fields_self[e_2].sel(f=freq, mode_index=mi0, drop=True)
+                h_self_1 = fields_self[h_1].sel(f=freq, mode_index=mi0, drop=True)
+                h_self_2 = fields_self[h_2].sel(f=freq, mode_index=mi0, drop=True)
+
+                for mi1 in mode_index_1:
+                    e_other_1 = fields_other[e_1].sel(f=freq, mode_index=mi1, drop=True)
+                    e_other_2 = fields_other[e_2].sel(f=freq, mode_index=mi1, drop=True)
+                    h_other_1 = fields_other[h_1].sel(f=freq, mode_index=mi1, drop=True)
+                    h_other_2 = fields_other[h_2].sel(f=freq, mode_index=mi1, drop=True)
+
+                    # Cross products of fields
+                    e_self_x_h_other = e_self_1 * h_other_2 - e_self_2 * h_other_1
+                    h_self_x_e_other = h_self_1 * e_other_2 - h_self_2 * e_other_1
+
+                    # Integrate over plane
+                    d_area = self._diff_area
+                    integrand = (e_self_x_h_other - h_self_x_e_other) * d_area
+                    dot[i, mi0, mi1] = 0.25 * integrand.sum(dim=d_area.dims)
+
+        coords = {"f": f, "mode_index_0": mode_index_0, "mode_index_1": mode_index_1}
+        return MixedModeDataArray(dot, coords=coords)
+
```

Comment:
This method seems unnecessarily long/complicated compared to how I would have gone about it, and also there's a lot of repeated code with the regular `dot`, which we always try to avoid. After thinking about it for a little bit, it seems like the main drawback is that we don't have `FieldData` methods to `interp` things and return another `FieldData` - instead you have to manually re-collect xarray DataArrays into a `FieldData`. We've gone back and forth on this a few times, whether we want to wrap these methods, and last time we decided not to, one reason being that we don't want `FieldData.monitor` to become stale w.r.t. the data stored in the data structure.

Still, it seems to me that a better approach would be something like:
- Get centered tangential fields of `self`
- Create a new `FieldData` with a monitor given by `other.monitor.copy(update=dict(colocate=True))`, and with fields given by the `other` fields `interp`-ed to the centers of `self`
- Iterate over the common frequencies and `i = range(self.mode_index.size)`, and do something likel `self._isel(mode_index=i).dot(colocated_other)`.

`_isel` is a method we do have implemented already for internal use. Maybe we could implement `_interp` too if it will simplify things.

Reply:
When other is a `ModeSolverMonitor`, `colocate` is defined to be always false, so I guess `other.monitor.copy(update=dict(colocate=True))` won't work.

I've refactored much of this code in the current commit, but not in this manner you mention yet.

Reply:
Yeah I feel like this whole handling maybe needs a rethinking, but also it feels like a low priority compared to other things we're working on now... The usual dilemma. I'll have a look when you think you have a final (for now) version though.